src/include/function/algorithm/parallel_utils.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include <atomic>
+#include <condition_variable>
+#include <mutex>
+
+#include "common/task_system/task_scheduler.h"
+#include "function/table_functions.h"
+#include "ife_morsel.h"
+#include "processor/operator/algorithm/algorithm_runner_worker.h"
+#include "processor/operator/call/in_query_call.h"
+#include "processor/operator/sink.h"
+
+using namespace kuzu::processor;
+
+namespace kuzu {
+namespace graph {
+
+class ParallelUtils {
```

Comment:
I am still confused about the functionality of ParallelUtils. Why does ParallelUtils take in an AlgorithmsRunnerWorker as an argument from outside? Ideally this class is something very internal to ParallelUtils, so should be constructed and maintained internally.

Also, the caller ShortestPath is calling this incrementTableFuncIdx to get ParallelUtils to run a particular function. That is you seem to be giving ParallelUtils a set of functions before it is constructed, and then call incrementTableFuncIdx and doParallel to run those functions. We can make ParallelUtils as a more standalone library that is not configured during construction. Instead ParallelUtils should expose functions that can be passed in functions. For example, if a caller wants ParallelUtils to call a particular function fx, it should call a function on ParallelUtils, e.g., ParalleUtils::doParallel(fx) and give fx to it as an argument. The ParalelUtils functions can be more specialized and take in Frontiers etc if those are the right primitives to have between ParallelUtils and Algorithms implementations.